backend/src/authentication/user.rs

```diff
@@ -0,0 +1,274 @@
+use axum_extra::extract::CookieJar;
+use serde::{Deserialize, Serialize};
+use sqlx::{FromRow, SqlitePool};
+
+use crate::{APIError, AppState};
+
+use super::{
+    error::AuthenticationError,
+    password::{hash_password, verify_password},
+    session::Sessions,
+    SESSION_COOKIE_NAME,
+};
+
+/// User object, corresponds to a row in the users table
+#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, Eq, Hash, FromRow)]
+pub struct User {
+    id: u32,
+    username: String,
+    password_hash: String,
+    updated_at: i64,
+    created_at: i64,
+}
+
+impl User {
+    pub fn id(&self) -> u32 {
+        self.id
+    }
+
+    pub fn username(&self) -> &str {
+        &self.username
+    }
+}
+
+/// Implement the FromRequestParts trait for User, this allows us to extract a User from a request
+/// using the user repository and the session cookie
+#[async_trait]
+impl<S> FromRequestParts<S> for User
+where
+    Users: FromRequestParts<S>,
+    S: Send + Sync,
+{
+    type Rejection = APIError;
+
+    async fn from_request_parts(parts: &mut Parts, state: &S) -> Result<Self, Self::Rejection> {
+        let Ok(users) = Users::from_request_parts(parts, state).await else {
+            return Err(AuthenticationError::UserNotFound.into());
+        };
+
+        let jar = CookieJar::from_headers(&parts.headers);
+        let Some(session_cookie) = jar.get(SESSION_COOKIE_NAME) else {
+            return Err(AuthenticationError::NoSessionCookie.into());
+        };
+
+        Ok(users.get_by_session_key(session_cookie.value()).await?)
+    }
+}
+
+pub struct Users(SqlitePool);
+
+impl Users {
+    #[cfg(test)]
+    pub fn new(pool: SqlitePool) -> Self {
+        Self(pool)
+    }
+
+    /// Authenticate a user by their username and password, returns a user instace on success or an error
+    pub async fn authenticate(
+        &self,
+        username: &str,
+        password: &str,
+    ) -> Result<User, AuthenticationError> {
+        let Some(user) = self.get_by_username(username).await? else {
+            return Err(AuthenticationError::UserNotFound);
+        };
+
+        if verify_password(password, &user.password_hash) {
+            Ok(user)
+        } else {
+            Err(AuthenticationError::InvalidPassword)
+        }
+    }
+
+    /// Get a user by their session key
+    pub async fn get_by_session_key(&self, session_key: &str) -> Result<User, AuthenticationError> {
+        let sessions = Sessions::new(self.0.clone());
+
+        // fetch the session from the database
+        let Some(session) = sessions.get_by_key(session_key).await? else {
+            return Err(AuthenticationError::SessionKeyNotFound);
+        };
+
+        // fetch the user from the database
+        let Some(user) = self.get_by_id(session.user_id()).await? else {
+            return Err(AuthenticationError::UserNotFound);
+        };
+
+        Ok(user)
+    }
+
+    /// Create a new user, save a Argon2id v19 hash of the password
+    #[allow(unused)]
```

Comment:
What does `#[allow(unused)]` mean?

Reply:
Cargo should not warn that this code is currently not called, since we need it in the future.